src/AzureRenderManager/WebApp/Code/Session/SessionTokenCache.cs

```diff
@@ -0,0 +1,104 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System.Security.Claims;
+using System.Text;
+using WebApp.Code.Extensions;
+
+namespace WebApp.Code.Session
+{
+    public class SessionTokenCache
+    {
+        private static readonly object FileLock = new object();
+        private readonly string _cacheKey;
+        private ClaimsPrincipal _claimsPrincipal;
+        private readonly IMemoryCache _memoryCache;
+        private TokenCache _cache = new TokenCache();
+
+        public SessionTokenCache(ClaimsPrincipal claimsPrincipal, IMemoryCache memoryCache)
+        {
+            _cacheKey = BuildCacheKey(claimsPrincipal);
+            _claimsPrincipal = claimsPrincipal;
+            _memoryCache = memoryCache;
+            Load();
+        }
+
+        private static string BuildCacheKey(ClaimsPrincipal claimsPrincipal)
+        {
+            // TODO: This needs to be updated in a multi-tenant env.
+            return $"UserId:{claimsPrincipal.Claims.GetObjectId()}";
+        }
+
+        public TokenCache GetCacheInstance()
+        {
+            _cache.BeforeAccess = BeforeAccessNotification;
+            _cache.AfterAccess = AfterAccessNotification;
+            Load();
+            return _cache;
+        }
+
+        public void SaveUserStateValue(string state)
+        {
+            lock (FileLock)
```

Comment:
The ASP.NET MemoryCache is safe to access from multiple threads, so lock shouldn't be needed.